### ISSUE TITLE
OSX BUILD

### ISSUE THREAD

**Issue by BPouet** [rbdrum2midi/rbdrum2midi#14]
Hi!
I would like to test the rbdrum2midi with RB2 GUITAR + RB3 GUITAR + RB2 DRUM + RB3 DRUM + RB3 KEYTAR. All Playstation.
I only code arduino so I understand a bit of the code but not used to it.
Would it be possible to release a build for osx?
If not, I will try to push on Cmake and follow your instructions.
Cheers!

**Comment by ssj71** [comment]
I believe it should work, the hard part is getting libraries that it will depend on such as libusb-1.0. That's probably available using homebrew, but I don't really deal with OSX much. Let me know what obstacles you hit and I'll try to help out where I can.

**Comment by BPouet** [comment]
I will give it a try. I see that in Cmake you can do a unix makefiles which should on OSX work but I think I will investigate homebrew. Thanks for the quick respond! Really appreciated it!

**Comment by BPouet** [comment]
Hi ssj71,

I have read and done some test, I shall need some guidance.

This is what I did 

`TO INSTALL HOMEBREW
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"

TO INSTALL libusb1.0 LIBRARY
brew install libusb

TO INSTALL LIBJACK
brew install jack

TO INSTALL libasound2
brew install portaudio


TO INSTALL CMAKE IN BREW
Brew install cmake

`
After that I try to run cmake to make a build.

**If I run cmake in the master folder i get** 

> build % cd /Users/bpouet/Downloads/rbdrum2midi-master 
 rbdrum2midi-master % mkdir build
 rbdrum2midi-master % cd build 
 build % cmake ..
CMake Warning (dev) in CMakeLists.txt:
  No project() command is present.  The top-level CMakeLists.txt file must
  contain a literal, direct call to the project() command.  Add a line of
  code such as

    project(ProjectName)

  near the top of the file, but after cmake_minimum_required().

  CMake is pretending there is a "project(Project)" command on the first
  line.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- The C compiler identification is AppleClang 11.0.3.11030032
-- The CXX compiler identification is AppleClang 11.0.3.11030032
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE) 
-- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE) 
-- Found Jack: /usr/local/lib/libjack.dylib  
CMake Error at /usr/local/Cellar/cmake/3.17.2/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:164 (message):
  Could NOT find ALSA (missing: ALSA_LIBRARY ALSA_INCLUDE_DIR)
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.17.2/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:445 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/Cellar/cmake/3.17.2/share/cmake/Modules/FindALSA.cmake:60 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  src/CMakeLists.txt:12 (find_package)


-- Configuring incomplete, errors occurred!
See also "/Users/bpouet/Downloads/rbdrum2midi-master/build/CMakeFiles/CMakeOutput.log".


The ALSA library is suppose to be in asound2 which is in portaudio I did at first.
There seems to be missing the name project in the CMAKELISTS but there is one.

**If I do a cmake in src folder I get** 

 ~ % cd /Users/bpouet/Downloads/rbdrum2midi-master/src 
 src % mkdir build
 src % cd build
 build % cmake ..
-- The C compiler identification is AppleClang 11.0.3.11030032
-- The CXX compiler identification is AppleClang 11.0.3.11030032
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE) 
CMake Error at CMakeLists.txt:11 (find_package):
  By not providing "FindJack.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Jack", but
  CMake did not find one.

  Could not find a package configuration file provided by "Jack" with any of
  the following names:

    JackConfig.cmake
    jack-config.cmake

  Add the installation prefix of "Jack" to CMAKE_PREFIX_PATH or set
  "Jack_DIR" to a directory containing one of the above files.  If "Jack"
  provides a separate development package or SDK, be sure it has been
  installed.


-- Configuring incomplete, errors occurred!
See also "/Users/bpouet/Downloads/rbdrum2midi-master/src/build/CMakeFiles/CMakeOutput.log".

This folder does exist so I think I got to be in the master folder. Create the build folder, go in that folder and do cmake parent folder ".." .

What could be my error?
My thanks again in advance!

**Comment by ssj71** [comment]
try first 'brew install pkg-config' then run it in the build directory. I think it should be identical in either one, but I'm not sure. Thanks for trying so much! Let me know the results after installing pkg-config and trying to run 'cmake ..' in the build directory again.

**Comment by BPouet** [comment]
I did install pkg-config and it seems to change nothing.
If I add 'project(rbdrum2midi)' in cmakelist and delete every line with ALSA, i am able to complete a build.
I come to understand that ALSA is our problem and the library is not for mac, PORTAUDIO is the solution.
Which mean the alsadriver.c would need to be convert to portaudiodriver.c with coding.

This project seems to have done it and can maybe help us to understand.
https://github.com/MycroftAI/mimic1/issues/13

Thanks for your advice!

**Comment by ssj71** [comment]
Yes that makes sense. ALSA is a linux specific driver. The proper thing is to have Cmake check which platorm it is building on and only build alsa if appropriate. As for a portaudio solution, I don't know that I'll have time to make such an effort, is using the JACK driver acceptable to you? Are you familiar with using JACK?

**Comment by BPouet** [comment]
Forget about portaudio. If Jack is compatible with your script than Iet's keep it simple.
The jack driver app doesn't seems to work on Mac 10.15 catalina (64 bits only) but qjackctl works.
It seems like it is an audio matrix, i am testing it.
Give me a day to play with it and check tutorials.
Thanks!

**Comment by BPouet** [comment]
Ok I think I figure it out with Jack.
I installed Jack V0.3.6
Use this link to setup the server
https://ccrma.stanford.edu/software/openmixer/manual/using_netjack/netjack-under-osx

How do you use jack with rbdrum2midi?
Compile a version without alsa?
I should be able to see rbdrum2midi into the connection of jack?
Thanks again!

**Comment by ssj71** [comment]
> How do you use jack with rbdrum2midi?

just run `rbdrum2midi -j` with the `-j` argument 

> Compile a version without alsa?

yes, this isn't something I've ever tried, but it shouldn't require huge changes to the code or build system. You mentioned commenting out some alsa stuff, have you been able to make a build?

> I should be able to see rbdrum2midi into the connection of jack?

yes, with qjackctl IIRC there are 2 different tabs in the connection window for midi, one is JACK midi and the other (at least on linux) is ALSA midi. Watch for that.

**Comment by BPouet** [comment]
I was able to compile and install by removing every line about alsa. Including &aseqq.
When I run rbdrum2midi -j with the dongle plug in I get :
 usb_claim_interface error -3 (busy=-6)
Error: MIDI driver setup failed.

I also noticed that the only driver that has midi_driver is alsa.
This means there may not be any midi for jack on mac.
I think this is related. I think I reached the end point.
Do you have any other idea I could push?
Thanks!

**Comment by ssj71** [comment]
Jack is a midi driver as well, there's no other reason for it to be part of this project. I think we are actually hitting USB permissions or other errors. I'm not sure if OSX is claiming the USB device as soon as you plug it in or something, but I'd search about USB device busy under libusb-1.0 on OSX. It may give some insight.

**Comment by YvanQu** [comment]
I managed to get the same error. Apparently libusb can't unload a kext in OSX, the way to do this would be use HIDAPI as a cross-platform wrapper that uses libusb for linux and whatever works on OSX for OSX. Right now it might be possible to manually find and unload the kext, but thats a serious PITA, especially if you want to keep using your drums for clone hero

**Comment by ssj71** [comment]
I unfortunately don't have access to an OSX computer so I am not able to support you in this. I'd be happy to merge a pull request though if someone can resolve this issue.

**Comment by YvanQu** [comment]
I tried messing around with it a bit, but it ended up being easier to tweak miditar hero. cheers!
